text/0000-api-evolution.md

```diff
@@ -0,0 +1,748 @@
+provide explicit contracts for their behavior via documentation, and behavior
+that is not part of this contract is permitted to change in minor
+revisions. (Remember: this RFC is about setting a *minimum* bar for when major
+version bumps are required.)
+
+This policy will likely require some revision over time, to become more explicit
+and perhaps lay out some best practices.
+
+### Signatures in type definitions
+
+#### Major change: tightening bounds.
+
+Adding new constraints on existing type parameters is a breaking change, since
+existing uses of the type definition can break. So the following is a major
+change:
+
+```rust
+// MAJOR CHANGE
+
+// Before
+struct Foo<A> { .. }
+
+// After
+struct Foo<A: Clone> { .. }
+```
+
+#### Minor change: loosening bounds.
+
+Loosening bounds, on the other hand, cannot break code because when you
+reference `Foo<A>`, you *do not learn anything about the bounds on `A`*. (This
+is why you have to repeat any relevant bounds in `impl` blocks for `Foo`, for
+example.) So the following is a minor change:
+
+```rust
+// MINOR CHANGE
+
+// Before
+struct Foo<A: Clone> { .. }
+
+// After
+struct Foo<A> { .. }
+```
+
+#### Minor change: adding defaulted type parameters.
+
+All existing references to a type/trait definition continue to compile and work
+correctly after a new defaulted type parameter is added. So the following is
+a minor change:
+
+```rust
+// MINOR CHANGE
+
+// Before
+struct Foo { .. }
+
+// After
+struct Foo<A = u8> { .. }
+```
+
+#### Minor change: generalizing to generics.
+
+A struct or enum field can change from a concrete type to a generic type
+parameter, provided that the change results in an identical type for all
+existing use cases. For example, the following change is permitted:
+
+```rust
+// MINOR CHANGE
+
+// Before
+struct Foo(pub u8);
+
+// After
+struct Foo<T = u8>(pub T);
+```
+
+because existing uses of `Foo` are shorthand for `Foo<u8>` which yields the
+identical field type.
+
+On the other hand, the following is not permitted:
+
+```rust
+// MAJOR CHANGE
+
+// Before
+struct Foo<T = u8>(pub T, pub u8);
+
+// After
+struct Foo<T = u8>(pub T, pub T);
+```
+
+since there may be existing uses of `Foo` with a non-default type parameter
+which would break as a result of the change.
+
+It's also permitted to change from a generic type to a more-generic one in a
+minor revision:
+
+```rust
+// MINOR CHANGE
+
+// Before
+struct Foo<T>(pub T, pub T);
+
+// After
+struct Foo<T, U = T>(pub T, pub U);
+```
+
+since, again, all existing uses of the type `Foo<T>` will yield the same field
+types as before.
+
+### Signatures in functions
+
+All of the changes mentioned below are considered major changes in the context
+of trait methods, since they can break implementors.
+
+#### Major change: adding new arguments.
+
+At the moment, Rust does not provide defaulted arguments, so any change in arity
+is a breaking change.
+
+#### Minor change: introducing a new type parameter.
```

Comment:
I wonder if we could allow hinting only a prefix of the generic type parameters, this would make situations like this where the new type parameter is inferrable always just work. I have at least one API where this would be useful myself (functions are usually called like either `foo::<X, _>()`, with the `_` essentially never needing to be specified).